kasa/cli.py

```diff
@@ -244,12 +278,22 @@ def _nop_echo(*args, **kwargs):
         return await ctx.invoke(discover, timeout=discovery_timeout)
 
     if type is not None:
-        device_type = DeviceType.from_value(type)
-        dev = await SmartDevice.connect(
-            host, credentials=credentials, device_type=device_type, timeout=timeout
+        dev = TYPE_TO_CLASS[type](host)
+        await dev.update()
+    elif device_family or encrypt_type:
+        connection_params = ConnectionParameters(
+            DeviceFamilyType(device_family),
+            EncryptType(encrypt_type),
+        )
+        dev = await Discover.connect_single(
+            host,
+            credentials=credentials,
+            timeout=timeout,
+            connection_params=connection_params,
+            try_discovery_on_error=False,
```

Comment:
I'm not sure if you saw my comment wrt storing all necessary information in connection parameters, so I'll repeat myself.

So what I had in mind was that we could store also rest of the information inside connection parameters, basically information like credentials, timeout options, host, used port, .. whatever is necessary to construct a device instance could be stored in this structure.

This way homeassistant (and others) can serialize the device settings in a very clean manner from an existing device instance, and re-create an instance by just doing something like `SmartDevice.connect(connection_params)`.

I think this would also allow us to simplify the internal structures for protocols and transports? Instead of passing the connection parameters separately, we can just pass the parameters object and let the internals to read the information from there.

What this also makes possible to have a single place for controlling sane defaults (where needed), e.g., the empty credentials object could be created when it's constructed and no creds are inputted.

Reply:
I'd missed the bit about storing host, credentials, timeout and port apologies.  I don't think this is such a good idea though if you mean to serialize it all:

- Host is stored in the config entry so would be duplicated.
- credentials are better stored centrally as all devices use the same ones.
- Timeout is likely to apply across all devices.
- Port does not seem to be used by HA and doesn't work with the new devices as they have 2 ports.

Unless you mean you want HA to be able to set **some** of these each time it supplies the params and only serialize the Protocol related ones?  i.e. a subclass called ProtocolParams that HA serializes.  TBH honest I'm not sure what this buys us and what happens if a user provides the same parameter as a kwarg and in the ConnectionParams.  I think it's simpler to restrict the ConnectionParams to the set needed to avoid discovery and keep it at that.

Reply:
Ok fair enough, I think I've been focused on using the ConnectionParams to get `Connect` working but I realise you're focused on reducing the proliferation of parameters in the method APIs.  That makes sense but I have a few clarifications:
- Do you agree HA should not need to persist the entire ConnectionParameters and only persist the things it's not already persisting elsewhere.  Storing host and credentials again in the config_entries doesn't seem right.
- Are we removing all the other parameters and replacing with ConnectionParams or do we keep some (`host` as least I think as it's been around since the beginning and no need to break the api)

Reply:
> Do you agree HA should not need to persist the entire ConnectionParameters and only persist the things it's not already persisting elsewhere. Storing host and credentials again in the config_entries doesn't seem right.

As we are already doing API breaking changes, I think it makes it just simpler to store the host and credentials in the same place. As a free bonus, any credentials will be bound to the specific config entry. What this means in practice is that if someone has devices that are not syncing the creds from the cloud (e.g., some are being blocked in separate vlans or whatever), those devices will still keep working as they were. 

Which made me wonder whether we should use the HA "store" to store the credentials at all, but rather keep it stored inside `hass.data[DOMAIN]`? This avoids storing the creds in plaintext at rest while still allowing adding multiple devices using a single authentication session per reboot.

For existing config entries, we should bump the config flow version and adapt the existing settings over to use the new format. If the `ConnectionParameters` defaults to using the legacy XOR on port 9999, the update flow could be as simple as something like this:
```
if config_entry.version == 1:
    dev = SmartDevice.connect(ConnectionParameters(**config_entry.data))
    hass.config_entries.async_update_entry(config_entry, data=dev.connection_parameters)
```

> Are we removing all the other parameters and replacing with ConnectionParams or do we keep some (host as least I think as it's been around since the beginning and no need to break the api)

If we decide to reuse the recently-added `connect()`, I doubt it makes a big difference. The host has been used in the ctors of `SmartDevice` derived classes and should remain so, but we are only concerned about the device factory for the time being. In the future, it might make sense to allow passing a connection parameters to allow defining those extra parameters like timeouts, ports, etc. but let's not worry about this for now.

So I think it's okay to break this API as the next release will be a breaking change anyway..  If we want to be really nice, maybe the function signature could be `connect(*, host=None, connection_parametrs=None)` to allow constructing device instances for known hosts using the legacy comms really easy?

Reply:
The original item of just passing ConnectionParameters to protocols & transports is now done for `SmartDevice.connect` so I'm resolving this conversation.  The subsequent discussion I'll quote at the bottom of the PR thread.